cmd/influx_inspect/buildtsi/util.go

```diff
@@ -0,0 +1,218 @@
+package buildtsi
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/influxdata/influxdb/models"
+	"github.com/influxdata/influxdb/tsdb"
+	"github.com/influxdata/influxdb/tsdb/engine/tsm1"
+	"github.com/influxdata/influxdb/tsdb/index/tsi1"
+	"go.uber.org/zap"
+)
+
+func IndexShard(sfile *tsdb.SeriesFile, dataDir, walDir string, maxLogFileSize int64, batchSize int, log *zap.Logger, verboseLogging bool) error {
+	log.Info("Rebuilding shard")
+
+	// Check if shard already has a TSI index.
+	indexPath := filepath.Join(dataDir, "index")
+	log.Info("Checking index path", zap.String("path", indexPath))
+	if _, err := os.Stat(indexPath); !os.IsNotExist(err) {
+		log.Info("tsi1 index already exists, skipping", zap.String("path", indexPath))
+		return nil
+	}
+
+	log.Info("Opening shard")
+
+	// Find shard files.
+	tsmPaths, err := collectTSMFiles(dataDir)
+	if err != nil {
+		return err
+	}
+	walPaths, err := collectWALFiles(walDir)
+	if err != nil {
+		return err
+	}
+
+	// Remove temporary index files if this is being re-run.
+	tmpPath := filepath.Join(dataDir, ".index")
+	log.Info("Cleaning up partial index from previous run, if any")
+	if err := os.RemoveAll(tmpPath); err != nil {
+		return err
+	}
+
+	// Open TSI index in temporary path.
+	tsiIndex := tsi1.NewIndex(sfile, "",
+		tsi1.WithPath(tmpPath),
+		tsi1.WithMaximumLogFileSize(maxLogFileSize),
+		tsi1.DisableFsync(),
+		// Each new series entry in a log file is ~12 bytes so this should
+		// roughly equate to one flush to the file for every batch.
+		tsi1.WithLogFileBufferSize(12*batchSize),
+	)
+
+	tsiIndex.WithLogger(log)
```

Comment:
Ah maybe this would cause problems if we added the `path` key above. I'm not sure how zap behaves if you add the same key multiple times...